### ISSUE TITLE
Add support for function call order expection

### ISSUE THREAD

**Issue by Kyle-Ye** [apple/swift-testing#297]
### Description

For async call or blocking call, we want to test the order the executing order. I'd like to propose an API set similar to [confirmation](https://swiftpackageindex.com/apple/swift-testing/main/documentation/testing/confirmation(_:expectedcount:fileid:filepath:line:column:_:))

### Expected behavior

```
func order2<R>(_ body: (Order2) async throw -> R)
func order3<R>(_ body: (Order3) async throw -> R)
func order4<R>(_ body: (Order4) async throw -> R)

@Test
func orderTest() {
  order3 { order in
    order.1()
    await Task {
      ...
      lock.wait()
      order.3()
    }
    await Task {
      ...
      order.2()
      lock.broadcast()
    }
  }
}

```

**Comment by grynspan** [comment]
Most scenarios that might need something like this are also solved by Swift concurrency, which is why we didn't implement an interface for it. For instance, if you expect functions A, B, and C to execute in that order, async/await means it's usually possible to force that to happen just by writing your calls to them in the right order.

Is there a scenario you're thinking of where that isn't possible today?

**Comment by Kyle-Ye** [comment]
> Most scenarios that might need something like this are also solved by Swift concurrency, which is why we didn't implement an interface for it. For instance, if you expect functions A, B, and C to execute in that order, async/await means it's usually possible to force that to happen just by writing your calls to them in the right order.
> 
> Is there a scenario you're thinking of where that isn't possible today?

Yes. For normal async/await Swift concurrency scenarios, there is no problem.

But for C-Style like POSIX lock API, it does not support very well. 

The Task and await code are just pseudo code to create 2 pthread. The core issue of the above example is to test the lock's behavior where Swift concurrency is not available.

However, such scenario should be very rare. Almost only a few low-level libraries that need to deal directly with the C language need such usage.

If upstream does not plan to support this feature, I totally understand the concern.

**Comment by grynspan** [comment]
swift-testing is, of course, a _Swift_ testing library. C code that behaves well when imported into Swift is testable too, but esoteric C behaviour is beyond the library's purview. That's not a _no_ to this feature request though: do you have a real-world example of C code that you need to use from Swift which you cannot turn into async/await code using continuations, tasks, task groups, etc.?

Something we could look at doing is using variadic generics and adding an overload of `confirmation()` _à la_:

```swift
await confirmation("first thing happened", "second thing happened", ..., strictlyOrdered: true) { first, second, ... in
}
```

This would help with the "arrow of death" pattern too.

**Comment by Kyle-Ye** [comment]
> swift-testing is, of course, a _Swift_ testing library. C code that behaves well when imported into Swift is testable too, but esoteric C behaviour is beyond the library's purview. That's not a _no_ to this feature request though: do you have a real-world example of C code that you need to use from Swift which you cannot turn into async/await code using continuations, tasks, task groups, etc.?

The example is the lock API I provide (NSLock, os_unfair_lock and so on). 

We have a wide usage of them in our ObjectiveC and Swift codebase. And we probably can convert them into Swift Concurrency but it's not on the plan.

> From https://developer.apple.com/videos/play/wwdc2021/10254/?time=1507
> Primitives like os_unfair_locks and NSLocks are also safe but caution is required when using them. Using a lock in synchronous code is safe when used for data synchronization around a tight, well-known critical section.

The core issue is we'd like to write test case for them easily. The current test implementation is a little tricky and unstable.

> Something we could look at doing is using variadic generics and adding an overload of confirmation() à la:

+1 for this idea/implementation. Look forward to seeing the landing PR.

**Comment by Kyle-Ye** [comment]
> So, for `NSLock` and `os_unfair_lock`/`OSAllocatedUnfairLock`, the primitive locking/unlocking functions are barred from use in async contexts and you should use the `withLock {}` API instead. This API has a built-in guarantee that the order of operations is correct, which eliminates the need for you to write a test that checks the order of operations. Hooray! 🥳

Sorry for the misunderstanding caused here. What we actually want to test here is something lower level than `withLock`.

The example is our cross-platform lock/condition implementation and we want to test the `wait` and `broadcast` API here.

> https://developer.apple.com/documentation/foundation/nscondition/1407950-wait
> https://developer.apple.com/documentation/foundation/nscondition/1415094-broadcast

**Comment by grynspan** [comment]
I would not recommend using `NSCondition` in Swift, especially in concurrent code. I know that's not a very satisfactory answer, but `NSCondition` is generally incompatible with Swift concurrency because it may cause a thread owned by Swift's concurrent thread pool to block indefinitely or even deadlock.

It is generally safe to assume that `NSCondition` _itself_ already has unit tests in Foundation that check that it behaves as designed, so you don't need to write your own unit tests just to check `NSCondition` itself.

Consider reimplementing your code to use Swift concurrency instead of a condition lock, and you may find that you (again!) don't need a dedicated interface in swift-testing to ensure your code operates correctly. 🙃

**Comment by Kyle-Ye** [comment]
> It is generally safe to assume that `NSCondition` _itself_ already has unit tests in Foundation that check that it behaves as designed, so you don't need to write your own unit tests just to check `NSCondition` itself.

Got it. But the problem is we are using our own implementation of `XXCondition` instead of Apple's `Foundation.NSCondition`. So I think we'd better add test case for it.

> I would not recommend using `NSCondition` in Swift, especially in concurrent code.

Got it. But that's a fairly old code in the codebase we cannot have a rewrite. I guess the best solution is just do not make the new Swift code to have a dependency on it.

Anyway thanks for your time on this issue. If upstream has plans to support such API later, we can keep this issue open, otherwise let's just close it.

**Comment by grynspan** [comment]
We can keep this issue open as there's still value in the proposed API.

**Comment by grynspan** [comment]
Tracked internally as rdar://133420399.

**Comment by juozas01589** [comment]
Higher level code could also benefit from the proposed API.

In my case, I have something like this:

```swift
struct BusinessLogic {
    let connection: ConnectionProtocol

    func run() async {
        await connection.send()
        await connection.receive()
    }
}
```

I would like to test the above code that:

- send and receive are both called
- they are called in the right order
- they are called at all, with a timeout

Currently, I can only do that with XCTest:

```swift
final class ConnectionMock: ConnectionProtocol {
    var onSend: (() async -> Void)?
    var onReceive: (() async -> Void)?

    func send() async {
        await onSend?()
    }

    func receive() async {
        await onReceive?()
    }
}

final class BusinessLogicTests: XCTestCase {
    func testRun() async {
        let didSend = expectation(description: "send called")
        let didReceive = expectation(description: "receive called")
        let connectionMock = ConnectionMock()
        connectionMock.onSend = { didSend.fulfill() }
        connectionMock.onReceive = { didReceive.fulfill() }

        let myLogic = BusinessLogic(connection: connectionMock)
        async let result = await myLogic.run()

        await fulfillment(of: [didSend, didReceive], timeout: 1, enforceOrder: true)
        await result
    }
}
```

**Comment by Kyle-Ye** [comment]
@mattmassicotte provides a solution to use `Event Streams`. https://www.massicotte.org/blog/testing-event-stream/
